### DISCUSSION TITLE
"debug test" with an already launched debugging session

### DISCUSSION THREAD

**Discussion by GLabat** [golang/vscode-go#1707] [Help]
Hi all. 

We have been using the VSCode Go extension (congrats for the integration) for several years now and for a couple of months, an issue impacts our debugging workflow.
We did not find any related issue on the subject and we are looking for advices here.

We use the following debugging pattern:
* start a debug session defined in `launch.json`
* use the mighty “debug test” to debug a specific test function which requires the previously started debug session

Whereas it used to work (not sure exactly when it stops working), we now get the following error message ‘Debug session has already been started’.

As a (painful) workaround, we use the following pattern:
* add a breakpoint on the first line of the test function
* click on “debug test”
* when the breakpoint is reached, start the debug session from launch.json
* continue debugging

Since this works, since the “compounds” section of `launch.json` also works, there seems to be no limitation in debugging several processes at the same time.

After a quick search, the culprit code seems to be located in
https://github.com/golang/vscode-go/blob/83aa2cb694262eb7f63af600a5c1958ca7435689/src/goMain.ts#L358
```go
vscode.commands.registerCommand('go.debug.cursor', (args) => {
			if (vscode.debug.activeDebugSession) {
				vscode.window.showErrorMessage('Debug session has already been started');
				return;
			}
			const goConfig = getGoConfig();
			testAtCursor(goConfig, 'debug', args);
		})
```

The final question is: why is the vscode.debug.activeDebugSession flag taken into account when locally debugging a specific test?

> A local build of the extension without the test on `vscode.debug.activeDebugSession` seems to be fine.

Should there be a better place to ask, please redirect me to it!
Thanks.

**Comment by hyangah** [comment]
That check was added as a workaround for https://github.com/golang/vscode-go/issues/109 (https://github.com/microsoft/vscode-go/issues/1383). Ability to use multiple debugging sessions is cool. I think more correct way to handle the original issue is to address only this specific case (debug session started with the test codelens).

**Reply by hyangah** [reply]
Filed https://github.com/golang/vscode-go/issues/1710

**Reply by GLabat** [reply]
Thanks for the analyse and issue @hyangah
